### DISCUSSION TITLE
PKCS5.PBKDF2 works much slower after v1.0.0

### DISCUSSION THREAD

**Discussion by sche** [krzyzanowskim/CryptoSwift#973] [General]
**Describe the bug**

After v1.0.0 the `PKCS5.PBKDF2` works much slower. It influences crypto-projects that rely on CryptoSwift.
E.g. 
HDWallet: https://github.com/essentiaone/HDWallet/blob/c1c5beb470d52673b89388b09d782345609d78f2/HDWalletKit/Mnemonic/Mnemonic.swift#L51
web3swift: https://github.com/matter-labs/web3swift/blob/e36e767eb2de26decb0c157aeef2d023a1175076/Sources/web3swift/KeystoreManager/BIP39.swift#L152 

Both libraries rely at the moment on CryptoSwift v1.0.0

Because of this, provided places in libs generate mnemonic seed several seconds with CryptoSwift v > 1.0.0

I tried to dig into to understand what causes it, and the bottleneck is here: https://github.com/krzyzanowskim/CryptoSwift/blob/af1b58fc569bfde777462349b9f7314b61762be0/Sources/CryptoSwift/SHA2.swift#L184

Benchmarking this loop in v1.0.0 compared to v1.2.0 demonstrates that in v1.0.0 it works much faster.

Here is a unit test to check it:
```swift
    func testPBKDF2Performance() {
        let passwordBytesHex = "73686f7020656c656374726963206c616464657220666f727761"
        let saltBytes = "6d6e656d6f6e6968"
        print("Start: \(Date())")
        _ = try! PKCS5.PBKDF2(password: Data(hex: passwordBytesHex).bytes,
                              salt: Data(hex: saltBytes).bytes,
                              iterations: 2048,
                              keyLength: 64,
                              variant: HMAC.Variant.sha512).calculate()
        print("End: \(Date())")
    }
```

Results for v1.0.0 on my mac with 2,4 GHz 8-Core Intel Core i9:
Start: 2020-10-07 13:11:01 +0000
End: 2020-10-07 13:11:01 +0000

Results for v1.2.0
Start: 2020-10-07 13:16:27 +0000
End: 2020-10-07 13:16:32 +0000

I tied to replace in 1.2.0 completely files `SHA2.swift` and `Utils.swift` with version from 1.0.0 but it did not help.

Probably this is an already known issue.

**Comment by krzyzanowskim** [comment]
Do you test Release build? I didn't notice this regression in performance tests hm.

**Comment by sche** [comment]
Yes, I tried different release builds.
So it works fine on your side (the example I provided)?

**Comment by grillr** [comment]
Same here! Version 1.0.0 is way faster than 1.3.2.

**Comment by krzyzanowskim** [comment]
do you use swift 5.3 on CryptoSwift 1.0? I'm yet to diagnose whether it's compiler or swift code regression

**Comment by grillr** [comment]
Yes, I use swift 5.3.

**Comment by krzyzanowskim** [comment]
This makes me think and I run test `PBKDFPerf` and that's what I see. That is for 65536 iterations

SHA1
v1.0.0 Avg 0.667
v1.3.0 Avg 0.796

SHA256
v1.0.0 Avg 0.445
v1.3.0 Avg 0.510

SHA512
v1.0.0 Avg 0.516
v1.3.0 Avg 0.640 

When I run 2048 iterations, as per @sche example, the result is
SHA256
v1.0.0 Avg 0.029
v1.3.0 Avg 0.032

Although there's a difference that is visible in numbers, it's not what I expected to see. If it takes you more than a second, that cany only means you're running unoptimized Swift (aka debug build). What I'm missing here?


My mac is 2,8 GHz Quad-Core Intel Core i7

**Comment by sche** [comment]
Hi @krzyzanowskim ,
I am testing PBKDF2 when CryptoSwift is imported via CocoaPods or SwiftPM.
I tried to run performance tests in CryptoSwift repo master branch, but I am getting a compilation error:
```
No 'swiftinterface' files found within '..../Build/Products/Test-iphonesimulator/CryptoSwift.framework/Modules/CryptoSwift.swiftmodule'.
Command PhaseScriptExecution failed with a nonzero exit code
``` 
> Although there's a difference that is visible in numbers, it's not what I expected to see. If it takes you more than a second, that cany only means you're running unoptimized Swift (aka debug build). What I'm missing here?

It works slow on optimized Swift for me when it is improted as a dependency. But only with versions > 1.0

Could you try to run the same performance test in a setup I described? (importing CryptoSwift with CocoaPods / SwiftPM)
Thanks!

**Comment by krzyzanowskim** [comment]
> I tried to run performance tests in CryptoSwift repo master branch, but I am getting a compilation error:

What is your Podfile, is this master branch or 1.3.2? what Xcode and CocoaPods version. Does Pod target uses Release for sure? There are so many configuration nuances with Cocoapods

>  (importing CryptoSwift with CocoaPods / SwiftPM)

I created SwiftPM sample package with your test code https://github.com/krzyzanowskim/CryptoSwift/issues/832#issue-716527818
[CryptoSwift-sample-test.zip](https://github.com/krzyzanowskim/CryptoSwift/files/5465076/CryptoSwift-sample-test.zip)

```
❯ swift run -c release          
Fetching https://github.com/krzyzanowskim/CryptoSwift.git
Cloning https://github.com/krzyzanowskim/CryptoSwift.git
Resolving https://github.com/krzyzanowskim/CryptoSwift.git at 1.3.2
[3/3] Linking CryptoSwift-sample-test
Duration: 0.020466089248657227 (seconds)
```

And the same but with Xcode

<img width="1344" alt="Screen Shot 2020-10-30 at 13 34 50" src="https://user-images.githubusercontent.com/758033/97705669-bf3fd880-1ab4-11eb-8d33-d3e8a3648794.png">
Duration: 0.020154953002929688 (seconds)

Can you run it and report back result?

**Comment by sche** [comment]
Hi @krzyzanowskim 

> What is your Podfile, is this master branch or 1.3.2? what Xcode and CocoaPods version. Does Pod target uses Release for sure? There are so many configuration nuances with Cocoapods

I checked out CryptoSwift master branch, opened CryptoSwift.xcodeproj and tried to run PBKDFPerf tests. 
I checked it one more time and it works now (I updated XCode recently and updated master branch, so problem solved here)

Your example on my mac:
```
% swift run -c release
Fetching https://github.com/krzyzanowskim/CryptoSwift.git
Cloning https://github.com/krzyzanowskim/CryptoSwift.git
Resolving https://github.com/krzyzanowskim/CryptoSwift.git at 1.3.2
[3/3] Linking CryptoSwift-sample-test
Duration: 0.022337913513183594 (seconds)
```

Here I tried to run same test in iOS app on iPhone 11 (14.1) simulator with CryptoSwift 1.3.2:
```
Start: 2020-11-06 17:29:53 +0000
End: 2020-11-06 17:29:57 +0000
```
[TestCryptoSwift.zip](https://github.com/krzyzanowskim/CryptoSwift/files/5502637/TestCryptoSwift.zip)

But now it is slow even if I try to run it with 1.0.0 version imported via SwiftPM 🤷‍♂️
Importing 1.0.0 via pod (v1.10.0) didn't work for me now... I am getting:
```
Undefined symbols for architecture x86_64:
  "(extension in CryptoSwift):Foundation.Data.init(hex: Swift.String) -> Foundation.Data", referenced from:
      TestCryptoSwiftPodTests.TestCryptoSwiftPodTests.testPBKDF2Performance() -> () in TestCryptoSwiftPodTests.o
  "(extension in CryptoSwift):Foundation.Data.bytes.getter : [Swift.UInt8]", referenced from:
      TestCryptoSwiftPodTests.TestCryptoSwiftPodTests.testPBKDF2Performance() -> () in TestCryptoSwiftPodTests.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
[TestCryptoSwiftPod.zip](https://github.com/krzyzanowskim/CryptoSwift/files/5502726/TestCryptoSwiftPod.zip)


But it works if I change version to 1.3.2 (which is still slow)

**Comment by krzyzanowskim** [comment]
> Here I tried to run same test in iOS app on iPhone 11 (14.1) simulator with CryptoSwift 1.3.2:

simulator speed is pretty much your mac speed, are you sure you didn't run debug build there?

> Importing 1.0.0 via pod (v1.10.0) didn't work for me now... I am getting:

some cocoapods setup leftovers gets in the way, either re-setup cocoapods, delete pods caches, derived data etc. happens with Xcode quite a lot.

**Comment by krzyzanowskim** [comment]
Starting v1.3.4 I distribute pre-compiled release version of the framework for Apple platforms. https://github.com/krzyzanowskim/CryptoSwift/issues/833#issuecomment-727293595
If you use SPM integration, you should use an Optimized build that most likely solves the issue you experience. This is the only way I can think of to make it happen at the moment.
